File: BasicLevel_C++/1014. 福尔摩斯的约会 (20).cpp
1014. 福尔摩斯的约会 (20)
大侦探福尔摩斯接到一张奇怪的字条：“我们约会吧！ 
3485djDkxh4hhGE 2984akDfkkkkggEdsb s&hgsfdk d&Hyscvnm”。
大侦探很快就明白了，字条上奇怪的乱码实际上就是约会的时间“星期四 14:04”，
因为前面两字符串中第1对相同的大写英文字母（大小写有区分）是第4个字母'D'，
代表星期四；第2对相同的字符是'E'，那是第5个英文字母，代表一天里的第14个
钟头（于是一天的0点到23点由数字0到9、以及大写字母A到N表示）；后面两字符串
第1对相同的英文字母's'出现在第4个位置（从0开始计数）上，代表第4分钟。现给定
两对字符串，请帮助福尔摩斯解码得到约会的时间。

输入格式：
输入在4行中分别给出4个非空、不包含空格、且长度不超过60的字符串。

输出格式：
在一行中输出约会的时间，格式为“DAY HH:MM”，其中“DAY”是某星期的3字符缩写，
即MON表示星期一，TUE表示星期二，WED表示星期三，THU表示星期四，FRI表示星期五，
SAT表示星期六，SUN表示星期日。题目输入保证每个测试存在唯一解。

输入样例：
3485djDkxh4hhGE
2984akDfkkkkggEdsb
s&hgsfdk
d&Hyscvnm
输出样例：
THU 14:04

分析：按照题目所给的方法找到相等的字符后判断即可，如果输出的时间不足2位数要在前面添0，即用%02d输出～

#include <iostream>
#include <cctype>
using namespace std;
int main() {
    string a, b, c, d;
    cin >> a >> b >> c >> d;
    char t[2];
    int pos, i = 0, j = 0;
    while(i < a.length() && i < b.length()) {
        if (a[i] == b[i] && (a[i] >= 'A' && a[i] <= 'G')) {
            t[0] = a[i];
            break;
        }
        i++;
    }
    i = i + 1;
    while (i < a.length() && i < b.length()) {
        if (a[i] == b[i] && ((a[i] >= 'A' && a[i] <= 'N') || isdigit(a[i]))) {
            t[1] = a[i];
            break;
        }
        i++;
    }
    while (j < c.length() && j < d.length()) {
        if (c[j] == d[j] && isalpha(c[j])) {
            pos = j;
            break;
        }
        j++;
    }
    string week[7] = {"MON ", "TUE ", "WED ", "THU ", "FRI ", "SAT ", "SUN "};
    int m = isdigit(t[1]) ? t[1] - '0' : t[1] - 'A' + 10;
    cout << week[t[0]-'A'];
    printf("%02d:%02d", m, pos);
    return 0;
}

File: BasicLevel_Golang/1014. 福尔摩斯的约会 (20).go
package main

import (
	"fmt"
	"unicode"
)

func main() {
	var (
		str  [4]string
		i, j int
		week = map[byte]string{
			'A': "MON",
			'B': "TUE",
			'C': "WED",
			'D': "THU",
			'E': "FRI",
			'F': "SAT",
			'G': "SUN",
		}
	)

	for i = 0; i < 4; i++ {
		fmt.Scanln(&str[i])
	}

	for i = 0; i < len(str[0]) && i < len(str[1]); i++ {
		if str[0][i] == str[1][i] && (str[0][i] <= 'G' && str[0][i] >= 'A') {
			fmt.Printf("%s ", week[str[0][i]])
			break
		}
	}
	i = i + 1
	for ; i < len(str[0]) && i < len(str[1]); i++ {
		if str[0][i] == str[1][i] {
			if unicode.IsDigit(rune(str[0][i])) {
				fmt.Printf("%02d:", str[0][i]-'0')
				break
			}
			if str[0][i] >= 'A' && str[0][i] <= 'N' {
				fmt.Printf("%02d:", str[0][i]-'A'+10)
				break
			}
		}
	}
	for j = 0; j < len(str[2]) && j < len(str[3]); j++ {
		if str[2][j] == str[3][j] && unicode.IsLetter(rune(str[2][j])) {
			fmt.Printf("%02d", j)
		}
	}
}
